### PR TITLE
SAK-50279 Samigo TA should only see assessments assignt to all site or assessment from the groups which does belong to

### PR DESCRIPTION
**Pull request by jesusmmp**
https://sakaiproject.atlassian.net/browse/SAK-50279

### CODE REVIEW

**Comment by ottenhoff** [comment]
Hi @jesusmmp , what is the performance hit from all of this extra work?

**Comment by jesusmmp** [comment]
> Hi @jesusmmp , what is the performance hit from all of this extra work?

I have not evaluated the performance, but within the new code there is nothing too complex.

**Review by bgarciaentornos — APPROVED** [review]
I'd say it doesn't add any heavy query, just looping through the groups.

So now it will behave the same way than assignments, I think that's nice.

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java [hunk 1]
```diff
@@ -724,8 +729,33 @@ public List<AssessmentFacade> getBasicInfoOfAllActiveAssessmentsByAgent(String o
 			if (questionSizeMap.get(a.getAssessmentBaseId()) != null) {
 				questionSize = ((Long) questionSizeMap.get(a.getAssessmentBaseId())).intValue();
 			}
-			AssessmentFacade f = new AssessmentFacade(a.getAssessmentBaseId(), a.getTitle(), a.getLastModifiedDate(), a.getStartDate(), a.getDueDate(), a.getReleaseTo(), releaseToGroups, lastModifiedBy, questionSize);
-			assessmentList.add(f);
+
+			if (releaseToGroups != null) {
+				try {
+					Site site = SiteService.getSite(siteAgentId);
+					Collection<Group> siteGroups = site.getGroupsWithMember(userId);
```
Comment (ottenhoff) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java:736]: are you sure all of this is cached? if you have 100 assessments, this seems like a huge amount of extra work

**Review by ottenhoff — CHANGES_REQUESTED** [review]
If you're the instructor, all of this extra work should be unncessary. Note that these calls look very expensive. Please, please test with a monitor on SQL calls..... you have some huge calls in here and repeating this logic for 100 assessments might be a huge amount of new SQL calls

**Comment by jesusmmp** [comment]
> If you're the instructor, all of this extra work should be unncessary. Note that these calls look very expensive. Please, please test with a monitor on SQL calls..... you have some huge calls in here and repeating this logic for 100 assessments might be a huge amount of new SQL calls

There are a lot of places where **site.getGroupsWithMember** is used
I don't think these pieces of code are a huge amount of extra work.

```
public Collection<Group> getGroupsWithMember(String userId)
	{
		Collection<Group> rv = new Vector<Group>();
		rv = getGroupsWithMembers(new String[] {userId});
		return rv;
	}
	
	/**
	 * {@inheritDoc}
	 */
	public Collection<Group> getGroupsWithMembers(String[] userIds)
	{
		Collection<Group> siteGroups = getGroups();
		ArrayList<String> siteGroupRefs = new ArrayList<String>(siteGroups.size());
		for ( Iterator it=siteGroups.iterator(); it.hasNext(); )
			siteGroupRefs.add( ((Group)it.next()).getReference() );
		
		List groups = authzGroupService.getAuthzUserGroupIds(siteGroupRefs, userIds[0]);
		Collection<Group> rv = new Vector<Group>();
		
		for (Iterator i = groups.iterator(); i.hasNext();)
		{
			Member m = null;
			Group g = getGroup( (String)i.next() );
			
			if ( g != null )
			{
				for (int j=0; j<userIds.length;j++)
				{
					m = g.getMember(userIds[j]);
					if ((m == null) || (!m.isActive()))
					{
						break;
					}
				}
				if ((m != null) && (m.isActive()))
				{
					rv.add(g);
				}
			}
		}

		return rv;
	}
```

**Comment by ottenhoff** [comment]
> There are a lot of places where site.getGroupsWithMember is used

Yes! But you call it inside a loop! So if there are 100 assessments, you are making the call 100 times.

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java (hunk 1 above)
Comment (ottenhoff) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java:736]: you should be doing work like this exactly once, not 100 times inside a loop

**Comment by jesusmmp** [comment]
> > There are a lot of places where site.getGroupsWithMember is used
> 
> Yes! But you call it inside a loop! So if there are 100 assessments, you are making the call 100 times.

Ok. I'll check to put these lines out of the loop. Thanks
```
Site site = SiteService.getSite(siteAgentId);
Collection<Group> siteGroups = site.getGroupsWithMember(userId);
```

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java [hunk 2]
```diff
@@ -711,10 +713,22 @@ public List<AssessmentFacade> getBasicInfoOfAllActiveAssessmentsByAgent(String o
...
+		String userId = AgentFacade.getAnonymousId();
+		GradingSectionAwareServiceAPI service = new GradingSectionAwareServiceImpl();
+		Site site = null;
+		Collection<Group> siteGroups = new ArrayList<>();
+		try {
+			site = SiteService.getSite(siteAgentId);
+			siteGroups = site.getGroupsWithMember(userId);
+		} catch (IdUnusedException ex) {
+			// no site found, just log a warning
+			log.warn("Unable to find a site with id ("+siteAgentId+") in order to get the enrollments, will return 0 enrollments");
```
Comment (ottenhoff) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java:726]: please use a distinct log message.

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java [hunk 3]
```diff
@@ -724,8 +738,25 @@ public List<AssessmentFacade> getBasicInfoOfAllActiveAssessmentsByAgent(String o
 			if (questionSizeMap.get(a.getAssessmentBaseId()) != null) {
 				questionSize = ((Long) questionSizeMap.get(a.getAssessmentBaseId())).intValue();
 			}
-			AssessmentFacade f = new AssessmentFacade(a.getAssessmentBaseId(), a.getTitle(), a.getLastModifiedDate(), a.getStartDate(), a.getDueDate(), a.getReleaseTo(), releaseToGroups, lastModifiedBy, questionSize);
-			assessmentList.add(f);
+
+			if (releaseToGroups != null) {
+				Map<String, String> groupIdsMap = siteGroups.stream()
+					.collect(Collectors.toMap(Group::getId, Group::getId));
```
Comment (ottenhoff) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java:744]: should this happen in the assessments loop or outside of the loop?

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/PublishedAssessmentFacadeQueries.java [hunk 4]
```diff
@@ -1430,30 +1432,58 @@ public List<PublishedAssessmentFacade> getBasicInfoOfAllPublishedAssessments2(St
 
 		List<PublishedAssessmentFacade> pubList = new ArrayList<>();
 		Map groupsForSite = null;
-		Map releaseToGroups;
+		Map releaseToGroups = new HashMap();
 		String lastModifiedBy = "";
 		AgentFacade agent = null;
-
+		Long assessmentId;
```
Comment (ottenhoff) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/PublishedAssessmentFacadeQueries.java:1438]:
```suggestion
		Long assessmentId = p.getPublishedAssessmentId();
```

Comment (jesusmmp) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/PublishedAssessmentFacadeQueries.java:1438]: p is defined on the loop
https://github.com/sakaiproject/sakai/pull/12713/files#diff-e7d70308680f6b42bafb2e164b377af7182711a351475f9bcf4e10bcee82c354R1450

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java (hunk 1 above)
Comment (jesusmmp) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java]: Solved

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java (hunk 2 above)
Comment (jesusmmp) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java]: Changed

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java (hunk 3 above)
Comment (jesusmmp) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/AssessmentFacadeQueries.java]: Changed outside the loop

**Comment by jesusmmp** [comment]
@ottenhoff This is ready to merge.

**Pull request merged**